pkg/controller/notification/snssubscription/controller.go

```diff
@@ -0,0 +1,243 @@
+	errUpdate                       = "failed to update the SNS Subscription"
+)
+
+// SetupSubscription adds a controller than reconciles SNSSubscription
+func SetupSubscription(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.SNSSubscriptionGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.SNSSubscription{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.SNSSubscriptionGroupVersionKind),
+			managed.WithExternalConnecter(&connector{
+				kube:        mgr.GetClient(),
+				newClientFn: sns.NewSubscriptionClient,
+				awsConfigFn: utils.RetrieveAwsConfigFromProvider,
+			}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithInitializers(),
+			managed.WithConnectionPublishers(),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube        client.Client
+	newClientFn func(*aws.Config) (sns.SubscriptionClient, error)
+	awsConfigFn func(context.Context, client.Reader, *corev1.ObjectReference) (*aws.Config, error)
+}
+
+func (conn *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+
+	cr, ok := mgd.(*v1alpha1.SNSSubscription)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+
+	awsconfig, err := conn.awsConfigFn(ctx, conn.kube, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := conn.newClientFn(awsconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, errClient)
+	}
+	return &external{c, conn.kube}, nil
+}
+
+type external struct {
+	client snsclient.SubscriptionClient
+	kube   client.Client
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.SNSSubscription)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	// AWS SNS resources are uniquely identified by an ARN that is returned
+	// on create time; we can't tell whether they exist unless we have recorded
+	// their ARN.
+	if !awsarn.IsARN(meta.GetExternalName(cr)) {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	// Get the list of Subscriptions with TopicARN
+	subscriptionList, err := e.client.ListSubscriptionsByTopicRequest(&awssns.ListSubscriptionsByTopicInput{
+		TopicArn: cr.Spec.ForProvider.TopicARN,
+	}).Send(ctx)
+	if err != nil {
+		return managed.ExternalObservation{},
+			errors.Wrap(resource.Ignore(sns.IsSubscriptionNotFound, err), errList)
+	}
+
+	// Filters the list of subscription with matching values in CR
+	sub, err := snsclient.GetSNSSubscription(subscriptionList, cr)
+	if err != nil {
+		return managed.ExternalObservation{},
+			errors.Wrap(resource.Ignore(sns.IsSubscriptionNotFound, err), errGetSubscription)
+	}
```

Comment:
It seems like this observe function does roughly two things:

1. Lists all subscriptions to the topic ARN of our subscription in order to determine whether our subscription exists.
1. Looks up the attributes of our subscription by its ARN in order to propagate its status.

According to https://docs.aws.amazon.com/sns/latest/api/API_GetSubscriptionAttributes.html the GetSubscriptionAttributesRequest will return HTTP 404 if the subscription does not exist. Could we therefore skip listing the subscriptions, and simply infer that the subscription did not exist if we called GetSubscriptionAttributesRequest and got a 404?

Reply:
Ah - I see that we use the returned `sub` to pass to `IsSNSSubscriptionUpToDate`. I don't think we need to do this though - see my comment on that function.